analyzer/protocol/ldap/ldap.spicy

```diff
@@ -0,0 +1,638 @@
+  # the best way to represent this as a string in a log.
+  # I've just left some of them as asn1::ASN1Message for now.
+
+  switch ( self.filterType ) {
+    FilterType::FILTER_AND     -> FILTER_AND:     asn1::ASN1Message(True, depth+1)
+                                                  &parse-from=self.filterBytes;
+    FilterType::FILTER_OR      -> FILTER_OR:      asn1::ASN1Message(True, depth+1)
+                                                  &parse-from=self.filterBytes;
+    FilterType::FILTER_NOT     -> FILTER_NOT:     SearchFilter(depth+1)
+                                                  &parse-from=self.filterBytes;
+    FilterType::FILTER_EQ      -> FILTER_EQ:      AttributeValueAssertion(depth+1)
+                                                  &parse-from=self.filterBytes;
+    FilterType::FILTER_SUBSTR  -> FILTER_SUBSTR:  SubstringFilter(depth+1)
+                                                  &parse-from=self.filterBytes;
+    FilterType::FILTER_GE      -> FILTER_GE:      AttributeValueAssertion(depth+1)
+                                                  &parse-from=self.filterBytes;
+    FilterType::FILTER_LE      -> FILTER_LE:      AttributeValueAssertion(depth+1)
+                                                  &parse-from=self.filterBytes;
+    FilterType::FILTER_PRESENT -> FILTER_PRESENT: asn1::ASN1OctetString(self.filter.head.len.len, False)
+                                                  &convert=$$.value.decode(hilti::Charset::ASCII)
+                                                  &parse-from=self.filterBytes;
+    FilterType::FILTER_APPROX  -> FILTER_APPROX:  AttributeValueAssertion(depth+1)
+                                                  &parse-from=self.filterBytes;
+    FilterType::FILTER_EXT     -> FILTER_EXT:     asn1::ASN1Message(True, depth+1)
+                                                  &parse-from=self.filterBytes;
+  };
+};
+
+type SearchRequest = unit(inout message: Message) {
+  var depth: uint32 = message.seq.body.seq.submessages[1].depth+1;
+
+  baseObject: asn1::ASN1Message(True, self.depth) &convert=$$.body.octetstring;
+  scope: asn1::ASN1Message(True, self.depth) &convert=cast<SearchScope>(cast<uint8>($$.body.num_value))
+                                             &default=SearchScope::NOT_SET;
+  deref: asn1::ASN1Message(True, self.depth) &convert=cast<SearchDerefAlias>(cast<uint8>($$.body.num_value))
+                                             &default=SearchDerefAlias::NOT_SET;
+  sizeLimit: asn1::ASN1Message(True, self.depth) &convert=$$.body.num_value &default=0;
+  timeLimit: asn1::ASN1Message(True, self.depth) &convert=$$.body.num_value &default=0;
+  typesOnly: asn1::ASN1Message(True, self.depth) &convert=$$.body.bool_value &default=False;
+  filter: SearchFilter(self.depth);
+  attributes: AttributeSelection(self.depth);
+
+  on %done {
+    message.obj = self.baseObject;
+    message.arg = "%s" % self.scope;
```

Comment:
Let's move these into `%done` hooks for `baseObject` and `scope`, respectively. Once we have done that we could probably hide the files after consuming their data.